File: bench/mapset.lua
-- The Lua twin of mapset.sl.
--
-- **Lua has neither a map type nor a set type: it has the table, and both are written with it.** So
-- the map is `m[k] = v` and the set is `s[k] = true`, which is the idiom every Lua program uses --
-- and it is a genuinely smaller amount of work than slate does, slate's `Map` hashing structurally
-- and consulting a class's own `==` where Lua compares keys by value or by reference.

local function run()
    local m, s, i = {}, {}, 0

    while i < 2000000 do
        m[i % 1000] = i
        s[i % 1000] = true
        i = i + 1
    end

    local total, k = 0, 0

    while k < 1000 do
        total = total + m[k]

        if s[k] then
            total = total + 1
        end

        k = k + 1
    end

    return total
end

print(run())

File: bench/mapset.php
<?php
// The PHP twin of mapset.sl.
//
// **PHP has neither a map type nor a set type: it has the array, and both are written with it** --
// `$m[$k] = $v` and `$s[$k] = true`, with `isset` as the membership test. That is Lua's situation,
// and it goes one step further: the keys here are the integers 0 to 999 arriving in order, so PHP
// keeps both arrays as plain vectors and never hashes anything. It is a genuinely smaller amount of
// work than slate's structurally hashed `Map` and `Set`, and it is what a PHP program gets. (The
// `ds` extension's `Map` and `Set` are not in a standard build.)

function run()
{
    $m = [];
    $s = [];
    $i = 0;

    while ($i < 2000000) {
        $m[$i % 1000] = $i;
        $s[$i % 1000] = true;
        $i = $i + 1;
    }

    $total = 0;
    $k = 0;

    while ($k < 1000) {
        $total = $total + $m[$k];

        if (isset($s[$k])) {
            $total = $total + 1;
        }

        $k = $k + 1;
    }

    return $total;
}

echo run(), "\n";
